### ISSUE TITLE
[Scala] Everything upper-case is now support.class

### ISSUE THREAD

**Issue by djspiewak** [sublimehq/Packages#549]
Wrong far more often than it is right.  I have a way to fix this, but it'll take somewhat longer.

**Comment by wbond** [comment]
Previously it was `entity.name.class` which was definitely incorrect. The new scope is the "usage" equivalent of the previous scope.

If you can provide some examples of where it is incorrect we can determine how they fit in with the TextMate scope conventions to see if there is a better fit.

**Comment by djspiewak** [comment]
I agree that `support.class` is slightly less wrong than `entity.name.class`.  So it's a step in the right direction.  This is mostly a ticket to track efforts to go further in that direction.

I can come up with a whole bunch of examples, but generally-speaking I would propose that we need to identify type expressions more explicitly.  Types can only appear in certain syntactically-delimited positions in Scala (e.g. within square brackets, following `:`, etc).  Outside of those positions, upper-case identifiers should probably be highlighted as `support.constant`, since that's going to be true 99.9% of the time in practical Scala (someone using upper-case for a variable is really just wrong, and it almost never happens).  For example:

```scala
val Foo = 42

Foo + 12
```

Thus, the first `Foo` would highlight as `entity.name.parameter`, as with any other constant variable declared with `val`, and the second `Foo` would highlight as `support.constant`.  A counter-example:

```scala
abc: Foo
```

This is a type ascription, and the `Foo` here should be highlighted as `support.type` or `support.class` (whatever makes more sense for sublime), and not as a constant.

I have a branch locally where I've been experimenting with this, and it seems that in nearly all cases, I can syntactically scope the type expressions, which means that my proposal is probably achievable.

**Comment by djspiewak** [comment]
Here's my current proposal on this one: https://gist.github.com/aec7d52de2efbf983a2378b3d0733da6  This test passes on my local branch.  Ignore the `def`, `val` etc stuff, since obviously we're tracking that in #550.

**Comment by wbond** [comment]
The new [Scope Naming documentation](http://www.sublimetext.com/docs/3/scope_naming.html) is probably worth spending some time reviewing.

Scoping `=>` feels wrong to me. Depending on its significance, I would expect `keyword.operator` or `punctuation.separator`. The first seems far more likely than the second. http://stackoverflow.com/a/7890032/230074 seems to indicate that it is a keyword also.

**Comment by djspiewak** [comment]
The StackOverflow answer you linked is inaccurate.  Or rather, it's deceptive.  `=>` is *slightly* magical, but it's not a keyword per se.  As I said, infix symbolic types are relatively easy to define in user-land.  Scalaz defines the `~>` operator:

```scala
val x: Foo ~> Bar
```

Shapeless defines `::`:

```scala
val xs: Int :: HNil
```

The only thing that's magical about `=>` is that it associates to the right.  Users can define right-associative infix types, but they must involve the `:` character (e.g. the `::` type example above is right-associative).

So in that sense, I think it's fair to highlight `=>` as a keyword, even though it's really sort of not.  Other symbolic infix types (like `~>`) really shouldn't be keywords and should almost certainly be highlighted as `support.type` or `support.class`.

**Comment by wbond** [comment]
I believe this was addressed in #563. Let me know if not.
